### PR TITLE
✨ Support for local cluster services DNS resolution

### LINKED ISSUE #505: Cross-namespace service DNS resolution in physical clusters
If a pod in namespace one references a service in another namespace, such as `some-service.two.svc.cluster.local`, how will that work in physical clusters where the namespace names are mapped and no longer retain their original names?

### LINKED ISSUE #1465: Same namespace fully qualified service DNS resolution in pcluster
**Is your feature request related to a problem? Please describe.**
Pods cannot connect to services when using fully qualified service names (e.g. `mysvc.mynamespace.svc.cluster.local`) because kcp does not preserve the original namespace (which is expected). 

This issue is related to https://github.com/kcp-dev/kcp/issues/505. The same problem exists for pods within the same namespace. Application developers can change their code to use non-qualified service names (e.g. `mysvc`) which is not ideal and might not be [optimal](https://pracucci.com/kubernetes-dns-resolution-ndots-options-and-why-it-may-affect-application-performances.html).

**Describe the solution you'd like**
Pods connecting to services using fully qualified service names (e.g. `mysvc.mynamespace.svc.cluster.local`) should succeed without any changes to the application code.

**Describe alternatives you've considered**
None.

**Additional context**

/cc @ncdc

### PR DESCRIPTION
**Pull request by lionelvillard**
## Summary

Configure synced deployments DNS Config to point to kcp DNS resolver mapping local namespaces to physical namespaces for cluster local services.
 
## Related issue(s)

Fixes #505 
Fixes #1465

### CODE REVIEW

**Comment by jmprusi** [comment]
@lionelvillard, thanks for the PR! 

A couple of comments:

I see that the PR doesn't take into account the workspace of the Pod that is making the request, so that could lead to conflicts in the namespace resolution, given the following scenario:

1. Pod A running in Namespace A belongs to workspace A
1. Service B in Namespace B belongs to workspace B
1. Service B in Namespace B belongs to workspace A

if POD A tries to resolve the service B in namespace B. Currently, the plugin can respond with a service from another Workspace (B)

To solve this, the plugin has to be aware of the Workspace of the querying POD and filter accordingly.

Also, wiring up the Syncer PODs IP to the dnsConfig means that if the syncer pod gets recreated (update, scale...), all the currently deployed pods/deployments will need to be resynced. Perhaps a service that fronts the syncer deployment with a cluster IP could be a better solution?

Thank you!

**Comment by lionelvillard** [comment]
> Also, wiring up the Syncer PODs IP to the dnsConfig means that if the syncer pod gets recreated (update, scale...), all the currently deployed pods/deployments will need to be resynced. Perhaps a service that fronts the syncer deployment with a cluster IP could be a better solution?

Yes that's a great idea, thanks! 

On that topic I feel a bit uneasy about packaging both the kcp dns proxy and the syncer together due to different scheduling requirements (the dns proxy is critical, the syncer is not). What's your take on this?

**Comment by jmprusi** [comment]
I agree. Coupling the DNS resolution to the syncer doesn't feel right... we were thinking of something along those lines:

* Deploying one CoreDNS pod for each workspace in a pcluster, with an updated map of the namespaces for the given workspace and pointing all the pods in all the workspace namespaces to their correct CoreDNS server. 

What is essential is to avoid extra watches/informers from being used and try to make the CoreDNS watch/hot-reload a configuration (configmap?) that is updated by the syncers on each namespace creation/deletion.

**Comment by lionelvillard** [comment]
> Deploying one CoreDNS pod for each workspace in a pcluster, with an updated map of the namespaces for the given workspace and pointing all the pods in all the workspace namespaces to their correct CoreDNS server.

I would much prefer this solution. It provides better workspaces isolation and can be easily enforced by using network policies. It also solves the source pod issue (we certainly don't want pod watch!) 

> What is essential is to avoid extra watches/informers from being used and try to make the CoreDNS watch/hot-reload a configuration (configmap?) that is updated by the syncers on each namespace creation/deletion.

There are various ways to do this. Using a configmap is simple but require a watch. Another way is to expose a management API (add/remove namespaces) which is quite easy to implement with Caddy.

**Comment by jmprusi** [comment]
> There are various ways to do this. Using a configmap is simple but require a watch. Another way is to expose a management API (add/remove namespaces) which is quite easy to implement with Cadd

Sorry, I meant avoiding extra watches against KCP, local watches should be fine! :)

**Comment by lionelvillard** [comment]
@jmprusi I implemented what we discussed above: 
- the kcp dns resolver now runs in its own pod and gets its configuration from the kcp-dns-nsmap configmap. 
- the syncer makes sure the kcp-dns-nsmap configmap is up-to-date
- the kubectl kcp plugin generates additional resources to deploy the kcp dns deployment. 

PTAL.

**Comment by lionelvillard** [comment]
Is it possible to get a /ok-to-test so I can check the e2e tests have been fixed? I will probably need several iterations. (it's working locally but I only run one configuration).

**Comment by lionelvillard** [comment]
@jmprusi Do you have time to look at this PR?

**Review by jmprusi — CHANGES_REQUESTED** [review]
Thanks! Worked perfectly!   ~~tried it locally and ran into some issues... I will add more reviews comments once I can resolve the local issue.~~

Not fully sure about 2 different images instead of using just one (syncer) launched with different flags, as that requires users to pass two different images...but I guess we can set a proper default image to avoid that?

**Comment by lionelvillard** [comment]
> Thanks! Worked perfectly! ~tried it locally and ran into some issues... I will add more reviews comments once I can resolve the local issue.~
> 

Great!


> Not fully sure about 2 different images instead of using just one (syncer) launched with different flags, as that requires users to pass two different images...but I guess we can set a proper default image to avoid that?

Having 2 different images is IMO cleaner (greater separation of concerns) and I agree the devex can be improved by setting a proper default image (when we get one).

> 
> Also, can we add some debug logging to the DNS plugin? (or document how to enable it)

Will do!

**Comment by lionelvillard** [comment]
@jmprusi I've updated the branch. PTAL! thx!

**Comment by lionelvillard** [comment]
@jmprusi is there anything left to be done for this PR, besides getting a ghcr.io/kcp-dev/kcp url?

**Comment by jmprusi** [comment]
I've run some tests locally and functionality LGTM, can you review @davidfestal and/or @sttts ?

**Comment by lionelvillard** [comment]
I just refactor a bit the CoreDNS plugin to closer match the CoreDNS [rewrite plugin](https://coredns.io/plugins/rewrite/).

File: docs/syncer.md
Comment (sttts) [inline docs/syncer.md:26]: why is this a different image? Can we merge them?

File: test/e2e/syncer/syncer_test.go
Comment (sttts) [reply test/e2e/syncer/syncer_test.go:453]: better: merge the two images. Having two is not good. Unneeded complexity. If it is hard to have two commands in one image, add a sub-command to the syncer which starts coredns instead.

File: docs/syncer.md
Comment (lionelvillard) [reply docs/syncer.md:26]: It is a different image because CoreDNS needs root access (see https://github.com/coredns/coredns/issues/1767#issuecomment-507213651) whereas the syncer does not. 

Instead of `--syncer-image` and `--dns-image` we could replace them by `--container-repository` and `--container-tag` and use fixed container image name.

File: test/e2e/syncer/syncer_test.go
Comment (lionelvillard) [reply test/e2e/syncer/syncer_test.go:453]: if security is not a strong concern I can merge the two images.

File: docs/syncer.md
Comment (sttts) [reply docs/syncer.md:26]: why does root access have anything to do with the image?

Comment (lionelvillard) [reply docs/syncer.md:26]: To run non-root containers, the image needs to be built with a non-root user and the k8s deployment needs to be configured to use the uid created in the image (see `securityContext.runAsUser`). Adding a non-root user in the image is not strictly required and could work if the syncer does not access the user directory.

Comment (lionelvillard) [reply docs/syncer.md:26]: There might be another solution based on container port mapping.

Comment (lionelvillard) [reply docs/syncer.md:26]: container port mapping works, problem solved. I'll use one image.

**Review by davidfestal — COMMENTED** [review]
Currently, a Syncer may sync resources form various user KCP workspaces.
More precisely from all workspaces which have bound to the compute APIExports related to the Syncer Synctarget.
In the future this situation would be the most usual one. Am I mistaken or is it right that this use-case is not supported ?
If I understand correctly this PR assumes that the only workspace resources synced by the syncer exist in the same KCP workspace that contains the SyncTarget.

Afaict, after discussing it with @jmprusi, supporting this mid-term scenario would mean either:
- installing a dedicated CoreDNS POD in a dedicated namespace for each couple SyncTarget / user KCP workspace. That would mean managing the creation of the CoreDNS POD from the Syncer itself... and a number of questions are raised by this....
- switching back to running the CoreDNS process as a sidecar in every user deployment POD (along with having the related Config map mounted as a volume on it), which possibly we don't want as well.

The second question I have is about adding NetworkPolicies in order to limit the network allowed communications to the CoreDNS POD(s) at their minimal. But I assume this one depends on what direction we would take on the first question.

Do you think we should open one (or 2) new issue(s) or an EPIC for these concerns and just merge this PR ?

cc @sttts

**Comment by lionelvillard** [comment]
> If I understand correctly this PR assumes that the only workspace resources synced by the syncer exist in the same KCP workspace that contains the SyncTarget.

Currently there is one DNS server per syncer. So when the Syncer is connected to multiple user workspaces, they all connect to the same DNS server. I haven't tested this use case yet so I'm most likely missing something.

Moving forward, I would like to merge this PR as, at least, it is useful for the simple one-to-one use case. If you could open follow up issues I'd be happy to take a look at them.

(side note: I'm working on fixing Stefan's concerns, almost done...)

**Comment by sttts** [comment]
> Moving forward, I would like to merge this PR as, at least, it is useful for the simple one-to-one use case

IMO @davidfestal 's concerns are fundamental. The described use-case by David is the normal use-case, not the other way around. We have to have a plan for that before merging. If we have that plan and decide to merge first, so be it. But this conceptual work IMO has to happen first.

If I remember right, Ben Bennet had discussed this in some doc of how to serve the right DNS records for each workspace, e.g. by source IP, or by having a proxy CoreDNS per workspace. @jmprusi can probably provide the link to that discussion.

**Comment by davidfestal** [comment]
> > Moving forward, I would like to merge this PR as, at least, it is useful for the simple one-to-one use case
> 
> IMO @davidfestal 's concerns are fundamental. The described use-case by David is the normal use-case, not the other way around. We have to have a plan for that before merging. If we have that plan and decide to merge first, so be it. But this conceptual work IMO has to happen first.
> 
So it is not excluded to merge this PR, even though the added feature would not be supported in the use-case of Location Workspaces.
But at least merging this **should not break** the ability of the Syncer to sync workload resources from several user workspaces, and have them up and running as it is today.


> If I remember right, Ben Bennet had discussed this in some doc of how to serve the right DNS records for each workspace, e.g. by source IP, or by having a proxy CoreDNS per workspace. @jmprusi can probably provide the link to that discussion.

@sttts @lionelvillard We should plan a design session to revisit the various options that were initially explorerd by @jmprusi and @knobunc

**Comment by jmprusi** [comment]
The document where we explored/discussed some options is: https://docs.google.com/document/d/1tzyYWMFlhgUL4IOTI6gjfBqTobsyXMjgFeqD7Uijn5g

> @sttts @lionelvillard We should plan a design session to revisit the various options that were initially explorerd by @jmprusi and @knobunc

Yes, please, let's arrange a design session.

**Comment by davidfestal** [comment]
So after our design session here is what I propose to do:
- Before merging this PR we should add a check so that the `DeploymentMutator` would replace the `DNSPolicy` and `DNSConfig` only when the deployment comes from the same upstream workspace as where the SyncTarget workspace.
By doing so we gate the scope of this PR to the only supported use-case, avoid breaking the behavior in the Location Workspace use-case.
- After the merge of this PR we should:
  - change the creation of the CoreDNS Deployment and associated resources from a static creation (in the `workload sync` CLI command) to have the syncer itself create it dynamically, one dpeloyment per user-workspace synced to this Syncer.
  - cleanup the CoreDNS Deployment when no more namespaces are related the corresponding user workspace.
This would be detailed in a dedicated issue.

File: pkg/syncer/spec/mutators/deployment.go
Comment (davidfestal) [inline pkg/syncer/spec/mutators/deployment.go:51]: > downstreamLogicalName logicalcluster.Name

I'd rather call it `syncTargetLogicalClusterName`. There are no logical clusters or workspaces **downstream** (== in the physical cluster).

**Comment by davidfestal** [comment]
@jmprusi Do you think you could test that now, after the last changes, this PR doesn't break the location workspace use-case anymore (introducing security risks on the physical cluster side) ?

File: pkg/syncer/spec/mutators/deployment.go
Comment (lionelvillard) [reply pkg/syncer/spec/mutators/deployment.go]: done

**Comment by jmprusi** [comment]
@lionelvillard can you rebase the commits and clean up the story?

**Comment by ncdc** [comment]
/hold
pending go.mod question. also continuing to review

File: cmd/syncer/cmd/dns.go
Comment (ncdc) [inline cmd/syncer/cmd/dns.go:25]: NewDNSCommand (xref https://github.com/golang/go/wiki/CodeReviewComments#initialisms)

File: pkg/cliplugins/workload/plugin/sync.go
Comment (ncdc) [inline pkg/cliplugins/workload/plugin/sync.go:573]: Global comment: DNS, or dns, but not Dns 😄 (https://github.com/golang/go/wiki/CodeReviewComments#initialisms)

File: pkg/dns/plugin/nsmap/name.go
Comment (ncdc) [inline pkg/dns/plugin/nsmap/name.go:24]: We don't need to define this interface here. You can embed a `*remapStringRewriter` in `nameRewriterResponseRule` and `valueRewriterResponseRule`.

https://github.com/golang/go/wiki/CodeReviewComments#interfaces

File: pkg/dns/plugin/nsmap/name.go
Comment (ncdc) [inline pkg/dns/plugin/nsmap/name.go:1]: Any chance you could add unit tests?

File: pkg/dns/plugin/nsmap/namespace.go
Comment (ncdc) [inline pkg/dns/plugin/nsmap/namespace.go:35]: Please have this take in a `ctx` so you can get the logger from it. We are not adding new non-contextual `klog` statements if we can avoid it.

File: pkg/dns/plugin/nsmap/setup.go
Comment (ncdc) [inline pkg/dns/plugin/nsmap/setup.go:35]: Please use a context with cancel. Add a logger to it. Then we can use it elsewhere.

File: pkg/dns/plugin/nsmap/config.go
Comment (ncdc) [inline pkg/dns/plugin/nsmap/config.go:39]:
```suggestion
func StartWatcher(ctx context.Context, callback OnUpdateFn) error {
```

**Review by ncdc — COMMENTED** [review]
I have to stop for the day. This is as far as I've made it - sorry it's not more. Will resume tomorrow.

File: pkg/dns/plugin/nsmap/name.go
Comment (lionelvillard) [reply pkg/dns/plugin/nsmap/name.go:24]: This is private code coming for CoreDNS, copied as is. I would rather leave it that way to not introduced bugs.

Comment (ncdc) [reply pkg/dns/plugin/nsmap/name.go:24]: If it's copied from another repo, please put it in the third_party dir

Comment (lionelvillard) [reply pkg/dns/plugin/nsmap/name.go:24]: Let me be more precise: this code is coming from the CoreDNS rewrite plugin and I've trimmed it down to just support exact rewriting. I didn't want to go to far in changing the original code but while I'm at it I can just trim it down even further and remove the interface

Comment (ncdc) [reply pkg/dns/plugin/nsmap/name.go:24]: Thanks. It still needs to be in third_party

File: pkg/syncer/dns/syncer.go
Comment (ncdc) [inline pkg/syncer/dns/syncer.go:50]: This must be a proper controller with a work queue and worker goroutines. We can't have event handler functions take lots of time (e.g. make network requests to update a configmap).

**Review by ncdc — CHANGES_REQUESTED** [review]
Still reviewing but posting what I have so far

File: pkg/dns/plugin/nsmap/name.go
Comment (ncdc) [reply pkg/dns/plugin/nsmap/name.go:24]: I would probably suggest including a link in a comment to the stackoverflow post, but that code does not need to be in third_party.

Everything coming from another repo must be identified as to where it originated from (ideally GitHub link w/specific commit), it must retain the original copyright header, and it needs to go in third_party.

File: pkg/syncer/spec/mutators/deployment.go
Comment (ncdc) [inline pkg/syncer/spec/mutators/deployment.go:214]: When would this be true? False? Why only do this when the deployment's logical cluster == the logical cluster where the SyncTarget lives?

File: pkg/syncer/syncer.go
Comment (ncdc) [inline pkg/syncer/syncer.go:212]: Would 1 second be ok?

File: pkg/syncer/syncer.go
Comment (ncdc) [inline pkg/syncer/syncer.go:215]: If err is not nil, I would store it in a variable declared outside the loop. Then, if the loop fails, include its details in the error returned on L221

File: test/e2e/virtual/syncer/virtualworkspace_test.go
Comment (ncdc) [inline test/e2e/virtual/syncer/virtualworkspace_test.go:658]:
```suggestion
						// Only need to install services in a non-logical cluster
```

File: test/e2e/syncer/syncer_test.go
Comment (ncdc) [inline test/e2e/syncer/syncer_test.go:75]:
```suggestion
				// Only need to install services in a logical cluster
```

File: test/e2e/syncer/syncer_test.go
Comment (ncdc) [inline test/e2e/syncer/syncer_test.go:521]:
```suggestion
				// Only need to install services in a logical cluster
```

File: test/e2e/reconciler/namespace/controller_test.go
Comment (ncdc) [inline test/e2e/reconciler/namespace/controller_test.go:99]:
```suggestion
							// Only need to install services in a logical cluster
```

File: test/e2e/reconciler/cluster/controller_test.go
Comment (ncdc) [inline test/e2e/reconciler/cluster/controller_test.go:193]:
```suggestion
						// Only need to install services in a non-logical cluster
```

File: pkg/syncer/spec/mutators/deployment.go
Comment (davidfestal) [reply pkg/syncer/spec/mutators/deployment.go:214]: That's because this PR provides only a limited implementation, and does not support the upcoming use-case where user workspaces and location workspaces are distinct.

See the following comment: https://github.com/kcp-dev/kcp/pull/1708#issuecomment-1243915910 and the comments before that led to this decision. The followup issue to support the more general case is here: https://github.com/kcp-dev/kcp/issues/1987

Comment (ncdc) [reply pkg/syncer/spec/mutators/deployment.go:214]: Could we please add a comment to the code itself with this information? Otherwise, it's not very clear.

File: pkg/syncer/dns/syncer.go
Comment (lionelvillard) [reply pkg/syncer/dns/syncer.go:42]: they are not available in this context.

Comment (ncdc) [reply pkg/syncer/dns/syncer.go:42]: Why not?

File: pkg/dns/plugin/nsmap/namespace.go
Comment (lionelvillard) [reply pkg/dns/plugin/nsmap/namespace.go:49]: actually it custom zone. `svc` is not configurable (see https://github.com/kubernetes/dns/blob/master/docs/specification.md)

I removed the TODO

File: pkg/syncer/spec/mutators/deployment.go
Comment (davidfestal) [reply pkg/syncer/spec/mutators/deployment.go:214]: @lionelvillard Could you add a TODO here ?

File: pkg/syncer/dns/syncer.go
Comment (lionelvillard) [reply pkg/syncer/dns/syncer.go]: I removed the DNS syncer which is now part of the downstream namespace controller. I didn't change the namespaceInformer and use the typed client to update/create ConfigMaps

File: pkg/dns/plugin/nsmap/namespace.go
Comment (lionelvillard) [reply pkg/dns/plugin/nsmap/namespace.go]: done.

File: pkg/dns/plugin/nsmap/setup.go
Comment (lionelvillard) [reply pkg/dns/plugin/nsmap/setup.go]: done.

File: pkg/syncer/dns/syncer.go
Comment (lionelvillard) [reply pkg/syncer/dns/syncer.go]: done. I moved the code to the downstream namespace controller.

File: pkg/syncer/spec/mutators/deployment.go
Comment (lionelvillard) [reply pkg/syncer/spec/mutators/deployment.go:215]: @davidfestal done!

**Comment by lionelvillard** [comment]
@davidfestal @jmprusi @ncdc this is ready for another round of reviews.

**Review by jmprusi — APPROVED** [review]

File: pkg/dns/plugin/nsmap/name.go
Comment (lionelvillard) [reply pkg/dns/plugin/nsmap/name.go]: this file has moved to third_party.

**Pull request merged**
